arangod/Aql/DocumentProducingNode.cpp

```diff
@@ -172,6 +172,13 @@ bool DocumentProducingNode::doCount() const noexcept {
   return _count && !hasFilter();
 }
 
+AsyncPrefetchEligibility DocumentProducingNode::canUseAsyncPrefetching()
+    const noexcept {
+  return (!hasFilter() || (_filter->isDeterministic() && !_filter->willUseV8()))
+             ? AsyncPrefetchEligibility::kEnableForNode
+             : AsyncPrefetchEligibility::kDisableForNode;
```

Comment:
Why do we have these constraints on the filter?

Reply:
The constraint for V8 is because the `Query` object only has a single V8 context, which it can enter and exit.
If multiple threads try to enter/exit the V8 context of the query concurrently, this would not be thread-safe.
The constraint for determinism is there because executing non-deterministic calculations could cause different query results if they are prefetched.
For example, a streaming query such as 
```
FOR i IN 1..1001
  LET x = ASSERT(i <= 1000)
  RETURN x
```
With the prefetching enabled, the query may fail before the first batch with the 1000 valid results is returned.